Add current metrics

martinpitt opened the pull request:
## Current state

![image](https://user-images.githubusercontent.com/200109/91710503-a5514a80-eb84-11ea-8b48-a816217aa223.png)

## @garrett's original design

![design](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/metrics/metrics-current.png)


## PR blockers
 - [x] Integration tests for network I/O

## Follow-ups
 - [x] Add disk I/O usage
 - [ ] Add load (blocked on https://github.com/cockpit-project/cockpit/pull/14510)
 - [ ] Add cgroups (top consumers of CPU and memory)

martinpitt commented:
I feel this is a good milestone to land now.

marusak reviewed (changes requested):
Looks great, firstly some design issue: (I am working with this screen)
![perfs](https://user-images.githubusercontent.com/12330670/91631570-d3932680-e9da-11ea-9346-d4fe623afda6.png)

1. The fact that the cards don't take the whole row feels rather ugly. But I guess it is what it is for now. I think that we don't scale cards to take the whole space on the overview page either. Also I think you want to add another card in the future.

2. Design has progress bar for each core of the CPU. I don't think it is necessary, just wanted to mention that.

3. For the memory, it feels wrong to have firstly the full capacity and then the used one. It reads like "32Gb out of 6.3... wait what? Ahh, 6.3 used out of 32, I see". Plus the dark part of the bar on the left shown the "used" part anyway.

4. Disk has the progress bar under description, but there is more space between description and related bar than between the bar and the description of the next location. Eye suggests that unrelated things are together. Also design suggests having description under the bar.

5. There is no full capacity of storage locations. I believe it should be shown, as having "42Gb free" is lacking some information value if you don't know from how much. (sure, there is the graph, but that is just to quickly see how the storage is doing, but should not expect that user will try to eyeball how many % is that and then do some maths for the storage capacity).

martinpitt commented:
> The fact that the cards don't take the whole row feels rather ugly.

Right, and they are sometimes too narrow as well. This needs an overhaul anyway. Also, we are going to add more cards.

> Design has progress bar for each core of the CPU.

Internal metrics don't expose individual cores. Also, it'd be fine for a small number of CPUs, but on larger servers with 32 or 96 CPUs it'd become really crowded. We can of course add a threshold there beyond which we show just one. But either way, this first needs a new (internal) metric channel in cockpit.

> Progress bar spacing

That's also an outstanding thing which needs CSS adjustment, right. I was thinking of waiting for Garrett next week, but I can try and fiddle around with it in this PR.

> There is no full capacity of storage locations

Good point, I'll add that and see how it looks.

martinpitt commented:
@marusak: I addressed your points 2 to 5, and also added current disk I/O. See updated screenshot.

@garrett, @marusak: For point 1 I plan to split CPU and memory cards into two, and show the top 5 CPU/memory consumers inside that card. IMHO that's more useful than a separate "processes" card, and it will conceptually match to the graphs if we have four cards.

One remaining consistency issue is that we show the amount of used RAM, but the amount of free space (per mount). Should we consistently show used or free for both?

marusak commented on test/check-application:
```diff
@@ -208,6 +208,26 @@ class TestApplication(testlib.MachineCase):
+        self.addCleanup(self.machine.execute, "umount /var/cockpittest")
+        self.assertLess(progressValue(".pf-c-progress[data-disk-usage-target='/var/cockpittest']"), 5)
+        free = b.text(".pf-c-progress[data-disk-usage-target='/var/cockpittest'] .pf-c-progress__status")
+        # anything between 40 and 50 MB
+        self.assertRegex(free, "^4\d\.\d MB free / 4\d\.\d MB total$")
```
40 to 50? This will get just to 49.9. Can it possibly be 50 exactly?

martinpitt replied:
No, there's always some internal file system overhead. `df --si -h` says total size is 47 MB.

marusak reviewed (approved):
Looks so much better now, thanks!
Went through the code and nothing stands out (except one question about tests). Those tests seem like they might become flaky (like expecting that the current traffic on `lo` is under MBs) but it all seems happy for now.

garrett commented:
> One remaining consistency issue is that we show the amount of used RAM, but the amount of free space (per mount). Should we consistently show used or free for both?

Showing free space for disks is definitely useful. If we want to make them both consistent, I'd suggest showing available RAM for the label instead. That's the most useful metric anyway, aside from swap (pressure).

Of course, just like with disks, there should be a tooltip on hover and focus to show extended details. (The extended info should probably be in ARIA as well, if the tooltip is not read.)

Also, I made the HTML mockup with PF4 widgets at the start, but they didn't work right, so I used a much simpler CSS grid that would work at all the various sizes instead. (With PF4 widgets for the layout, to best use the space meant fighting and overriding PF4.) Of course, the widgets inside the containers should still be PF4 — it's just the layout that really probably shouldn't be in this case.

garrett commented:
Additionally, the mockups have grey for the history area to show something should go there, not that the graphs should have a grey background. I can fix this though, when I'm working on the mockup. :wink:

martinpitt commented:
@marusak: I replaced "used" memory with "available" -- PTAL?

marusak reviewed (approved):
Thanks!
